### DISCUSSION TITLE
Callback 함수가 무엇인가요?

### DISCUSSION THREAD

**Discussion by BaekYuri** [wanted-9th-3team/frontend-study#64] [프론트 기술 질문 공간]
어떤 특징이 있는지, 왜 사용하는지 알려주세요.
어떠한 단점이 있는 것으로 알고 있는데, 이를 해결하기 위한 방안으로 나온 문법에 대해서도 설명해주세요.

**Comment by michoball** [comment]
Callback 함수는 함수안에 파라미터로 함수를 받아서 실행하는 구조를 가져 특정 함수가 실행되고나서 순차적으로 다른 함수를 실행시키는데 사용됩니다.  addEventListener나 setTimeout 과 같은 함수에서 볼 수 있습니다.  

순차적으로 함수들을 처리해주고 다양한 함수를 파라미터로 받아서 일을 처리 할 수 있기 때문에 함수를 유연하게 해주지만 자칫 잘 못하면 콜백함수들을 연속적으로 계속 사용해 콜백에 콜백에 콜백에 콜백함수를 사용하게 되는 등 일명 콜백지옥에 빠질 수도 있습니다.  

js에서는 이를 해결하기 위해 promise와 같은 문법을 만들었습니다. 하지만 promise 또한 then , catch 과 같은 함수로 resolve나 reject 된 상태에 따른 처리를 해야하는 불편함이 있었고 그래서 async await 을 만들었습니다.

**Comment by WooGie911** [comment]
콜백 함수란 다른 함수의 인자로써 이용되는 함수, 혹은 어떤 이벤트에 의해 호출되어지는 함수입니다.

주로 비동기(Asynchronous) 처리 방식의 문제점을 해결하기 위해 사용합니다.

하지만 콜백함수를 너무 남용하면 코드의 가독성이 떨어져 일명 콜백 지옥에 빠지고 에러 처리가 어렵습니다.

이를 해결하기 위해 promise가 등장하였고 then , catch를 함께 사용해야 한다는 promise의 문제점을 보안하기 위해 async/await가 등장하였고 try/catch로 예외처리가 가능합니다.

**Comment by javascriptbiggosoo** [comment]
비동기 작업을 할 때 콜백을 이용해 콜을 반복해서 받고 처리하는 경우에 생기는 유명한 콜백지옥이라는 문제가 있습니다.
```jsx
setTimeout(() => {
  console.log(`1 second passed`);
  setTimeout(() => {
    console.log(`2 second passed`);
    setTimeout(() => {
      console.log(`3 second passed`);
      setTimeout(() => {
        console.log(`4 second passed`);
      }, 1000);
    }, 1000);
  }, 1000);
}, 1000);
```
일단 가독성이 엄청 구려지는데, 그 말은 코드 형태 자체가 추후 새로운 버그를 유발할 확률을 높일 수 있다는 말이 됩니다.

그래서 탄생한 것이 여러 블록을 중첩하지 않고 한줄로 이어나가게끔 해준 Promise.
그리고 이 Promise의 번거로운 체이닝까지 해결한 async/await가 있습니다.
